### PR TITLE
ARROW-6084: [Python] Support LargeList

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

File: python/pyarrow/tests/strategies.py [hunk 1]
```diff
@@ -104,7 +104,10 @@ def fields(type_strategy=primitive_types):
 
 
 def list_types(item_strategy=primitive_types):
-    return st.builds(pa.list_, item_strategy)
+    return (
+        st.builds(pa.list_, item_strategy) |
+        st.builds(pa.large_list, item_strategy)
+        )
```
Comment (pitrou) [inline python/pyarrow/tests/strategies.py]: @kszucs Does this look right?

File: python/pyarrow/scalar.pxi [hunk 2]
```diff
@@ -586,6 +586,57 @@ cdef class ListValue(ArrayValue):
         return result
 
 
+cdef class LargeListValue(ArrayValue):
```
Comment (jorisvandenbossche) [inline python/pyarrow/scalar.pxi]: It's not possible to have this subclass ListValue to reduce code duplication? (getitem, iter, as_py look the same)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good to me

**Comment by jorisvandenbossche** [comment]
Again, while trying it out this branch, I noticed some oddities. But now they are certainly not related to the code in this PR, as they also hold for non-large ListType. 

When creating an array from offsets and values in python, there is no validation of the offsets that it starts with 0 and ends with the length of the array (but is that required? the docs seem to indicate that: https://github.com/apache/arrow/blob/master/docs/source/format/Layout.rst#list-type ("The first value in the offsets array is 0, and the last element is the length of the values array."). 
The array you get "seems" ok (the repr), but on conversion to python or flattened arrays, things go wrong:

```
In [61]: a = pa.ListArray.from_arrays([1,3,10], np.arange(5)) 

In [62]: a
Out[62]: 
<pyarrow.lib.ListArray object at 0x7fdd9c468678>
[
  [
    1,
    2
  ],
  [
    3,
    4
  ]
]

In [63]: a.flatten()
Out[63]: 
<pyarrow.lib.Int64Array object at 0x7fdd9cbfe9e8>
[
  0,   # <--- includes the 0
  1,
  2,
  3,
  4
]

In [64]: a.to_pylist()
Out[64]: [[1, 2], [3, 4, 1121, 1, 64, 93969433636432, 13]]  # <--includes more elements as garbage
```

I understand that in C++ the main constructors are not safe, and as the caller you need to ensure that the data is correct or call a safe (slower) constructor. But do we want to use the unsafe / fast constructors without validation in Python as default as well?

**Comment by jorisvandenbossche** [comment]
Ah, I see that in Python a validation function is exposed, and this actually raises:

```
In [65]: a.validate()                                                                                                                                                                                              
---------------------------------------------------------------------------
ArrowInvalid                              Traceback (most recent call last)
<ipython-input-65-b377de7e0b22> in <module>
----> 1 a.validate()

~/scipy/repos/arrow/python/pyarrow/array.pxi in pyarrow.lib.Array.validate()

~/scipy/repos/arrow/python/pyarrow/error.pxi in pyarrow.lib.check_status()

ArrowInvalid: Final offset invariant not equal to values length: 10!=5
```

Do we want to call that by default in the creation method? 
A quick search seems to indicate that `pa.Array.from_buffers` does this, but other `from_arrays` method don't seem to explicitly do this. But eg in DictionaryArray.from_arrays, if you pass values that is too short for the max of the indices, you do get an error.

**Comment by pitrou** [comment]
On the C++ side indeed, calling validation for each constructor would be expensive. We can adopt a different strategy for Python. Can you open a JIRA about that?

**Comment by jorisvandenbossche** [comment]
Opened https://issues.apache.org/jira/browse/ARROW-6132

File: python/pyarrow/scalar.pxi (hunk 2 above)
Comment (pitrou) [reply python/pyarrow/scalar.pxi]: The problem is that the C++ types are different (e.g. `ListArray` vs. `LargeListArray`), and those need to be compile-time constants for Cython.

**Comment by pitrou** [comment]
@wesm Do you want to take a look here or can I merge as-is? (assuming the R AppVeyor failure isn't related)

**Comment by pitrou** [comment]
Rebased.

**Comment by wesm** [comment]
Taking a look

File: python/pyarrow/scalar.pxi (hunk 2 above)
Comment (wesm) [reply python/pyarrow/scalar.pxi]: Code like this makes me wish for some kind of macro system in Cython.

**Review by wesm — APPROVED** [review]
+1

**Pull request closed without merging**